staging/src/k8s.io/apiserver/pkg/tracing/config_test.go

```diff
@@ -0,0 +1,240 @@
+		createFile     bool
+		expectedResult *apiserver.TracingConfiguration
+		expectedError  *string
+	}{
+		{
+			name:           "empty",
+			createFile:     true,
+			contents:       ``,
+			expectedResult: &apiserver.TracingConfiguration{},
+			expectedError:  nil,
+		},
+		{
+			name:           "absent",
+			createFile:     false,
+			contents:       ``,
+			expectedResult: nil,
+			expectedError:  strptr("unable to read tracing configuration from \"test-tracing-config-absent\": open test-tracing-config-absent: no such file or directory"),
+		},
+		{
+			name:       "v1alpha1",
+			createFile: true,
+			contents: `
+apiVersion: apiserver.config.k8s.io/v1alpha1
+kind: TracingConfiguration
+endpoint: localhost:4317
+samplingRatePerMillion: 12345
+`,
+			expectedResult: &apiserver.TracingConfiguration{
+				TypeMeta: metav1.TypeMeta{
+					Kind:       "",
+					APIVersion: "",
+				},
+				Endpoint:               &localhost,
+				SamplingRatePerMillion: &samplingRate,
+			},
+			expectedError: nil,
+		},
+		{
+			name:       "wrong_type",
+			createFile: true,
+			contents: `
+apiVersion: apps/v1
+kind: DaemonSet
+metadata:
+  name: agent
+spec:
+  selector:
+    matchLabels:
+      k8s-app: agent
+  template:
+    metadata:
+      labels:
+        k8s-app: agent
+    spec:
+      containers:
+        - image: k8s.gcr.io/busybox
+          name: agent
+`,
+			expectedResult: nil,
+			expectedError:  strptr("unable to decode tracing configuration data: no kind \"DaemonSet\" is registered for version \"apps/v1\" in scheme"),
+		},
+	}
+
+	for _, tc := range testcases {
+		t.Run(tc.name, func(t *testing.T) {
+			proxyConfig := fmt.Sprintf("test-tracing-config-%s", tc.name)
+			if tc.createFile {
+				f, err := ioutil.TempFile("", proxyConfig)
+				if err != nil {
+					t.Fatal(err)
+				}
+				defer os.Remove(f.Name())
+				if err := ioutil.WriteFile(f.Name(), []byte(tc.contents), os.FileMode(0755)); err != nil {
+					t.Fatal(err)
+				}
+				proxyConfig = f.Name()
+			}
+			config, err := ReadTracingConfiguration(proxyConfig)
+			if err == nil && tc.expectedError != nil {
+				t.Errorf("calling ReadTracingConfiguration expected error: %s, did not get it", *tc.expectedError)
+			}
+			if err != nil && tc.expectedError == nil {
+				t.Errorf("unexpected error calling ReadTracingConfiguration got: %#v", err)
+			}
+			if err != nil && tc.expectedError != nil && !strings.HasPrefix(err.Error(), *tc.expectedError) {
+				t.Errorf("calling ReadTracingConfiguration expected error: %s, got %#v", *tc.expectedError, err)
+			}
+			if !reflect.DeepEqual(config, tc.expectedResult) {
+				t.Errorf("problem with configuration returned from ReadTracingConfiguration expected: %#v, got: %#v", tc.expectedResult, config)
+			}
+		})
+	}
+}
+
+func TestValidateTracingConfiguration(t *testing.T) {
+	samplingRate := int32(12378)
+	negativeRate := int32(-1)
+	tooHighRate := int32(1000001)
+	validEndpoint := "localhost:4317"
+	dnsEndpoint := "dns:https://google.com:4317"
```

Comment:
Is `dns:https://` correct here? These [gRPC docs](https://github.com/grpc/grpc/blob/master/doc/naming.md) show `dns://`, I think?

Its test case below has `expectError: false`.

Reply:
Changed to `dns://`, which is correct.